File: ios/App/OpenPPP2/Runtime/RuntimeStore.swift
import Combine
import Foundation

@MainActor
public final class RuntimeStore: ObservableObject {
    @Published public private(set) var state: RuntimeSnapshot

    public init(
        initial: RuntimeSnapshot = RuntimeSnapshot(
            generation: 0,
            monotonicMs: 0,
            phase: .idle,
            capabilities: RuntimeSnapshot.bundledCapabilities
        )
    ) {
        state = initial
    }

    @discardableResult
    public func apply(_ incoming: RuntimeSnapshot) -> Bool {
        if incoming.generation < state.generation {
            return false
        }
        if incoming.generation == state.generation,
           incoming.monotonicMs <= state.monotonicMs {
            return false
        }
        state = incoming
        return true
    }

    public func markUnknown() {
        guard state.phase != .unknown else { return }
        state.phase = .unknown
    }

    @discardableResult
    public func applyUnknown(generation: UInt64, monotonicMs: UInt64) -> Bool {
        if generation < state.generation ||
            (generation == state.generation && monotonicMs <= state.monotonicMs) {
            return false
        }
        state = RuntimeSnapshot(
            generation: generation,
            monotonicMs: monotonicMs,
            phase: .unknown
        )
        return true
    }
}

File: android/lib/runtime/runtime_store.dart
import 'package:flutter/foundation.dart';

import 'runtime_snapshot.dart';

class RuntimeStore extends ChangeNotifier {
  RuntimeStore({RuntimeSnapshot? initial})
      : _state = initial ??
            const RuntimeSnapshot(
              generation: 0,
              monotonicMs: 0,
              phase: RuntimePhase.idle,
            );

  RuntimeSnapshot _state;
  RuntimeSnapshot get state => _state;

  bool apply(RuntimeSnapshot incoming) {
    if (incoming.generation < _state.generation) {
      return false;
    }
    if (incoming.generation == _state.generation &&
        incoming.monotonicMs <= _state.monotonicMs) {
      return false;
    }
    _state = incoming;
    notifyListeners();
    return true;
  }
}
